### PR TITLE
increase code readabiliy, preallocate vectors when possible

### PR DESCRIPTION
**Pull request by matthiasstraka**
The database code can be reduced in several places, especially row unique_ptrs only make sense to live inside the body of if/while statements.
Result vectors can and should be preallocated when possible using `getNumRows()` to avoid repeated reallocation.
In `SQLDatabase::getPathIDs`, the `LIMIT 1` clause has no effect since `WHERE` selects a unique id anyway.

### CODE REVIEW

File: src/database/sql_database.cc [hunk 1]
```diff
@@ -618,15 +616,18 @@ std::vector<std::shared_ptr<SQLDatabase::AddUpdateTable>> SQLDatabase::_addUpdat
...
     }
 
     if (obj->getParentID() == INVALID_OBJECT_ID) {
         throw_std_runtime_error("Tried to create or update an object {} with an illegal parent id {}", obj->getLocation().c_str(), obj->getParentID());
     }
 
     cdsObjectSql["parent_id"] = fmt::to_string(obj->getParentID());
-    returnVal.push_back(std::make_shared<AddUpdateTable>(CDS_OBJECT_TABLE, cdsObjectSql, op));
+
+    std::vector<std::shared_ptr<SQLDatabase::AddUpdateTable>> returnVal = {
```
Comment (neheb) [inline src/database/sql_database.cc:628]: = { not needed.

Alternatively, you could rewrite to

```
auto returnVal = std::vector {

CTAD will take care of the template argument.
```

File: src/database/sql_database.cc [hunk 2]
```diff
@@ -1471,16 +1459,16 @@ std::string SQLDatabase::incrementUpdateIDs(const std::unordered_set<int>& ids)
...
+    rows.reserve(res->getNumRows());
+    while (auto row = res->nextRow()) {
         std::ostringstream s;
         s << row->col(0) << ',' << row->col(1);
         rows.emplace_back(s.str());
     }
 
     if (rows.empty())
-        return "";
+        return {};
```
Comment (neheb) [inline src/database/sql_database.cc:1471]: I'd add a newline.

Comment (matthiasstraka) [reply src/database/sql_database.cc:1470]: Why a newline? We want to return an empty string (i.e. default constructed)

Comment (neheb) [reply src/database/sql_database.cc:1470]: No I mean between both return statements.

Comment (matthiasstraka) [reply src/database/sql_database.cc:1470]: Am I seeing the wrong line which you are talking about? There in only one modified return.

Comment (neheb) [reply src/database/sql_database.cc:1470]:
```
        return {};
<-------newline here
    return join(rows, ",");
```

Comment (matthiasstraka) [reply src/database/sql_database.cc:1470]: I only saw that after switching to full file view. Yes, a new line would look better. I can improve this tomorrow.

**Comment by KarlStraussberger** [comment]
Thank you for the contribution.

Please be aware that there is a PR pending for database code. So some of your changes will have to be reworked.

File: src/database/sql_database.cc [hunk 3]
```diff
@@ -855,12 +851,9 @@ std::vector<std::shared_ptr<CdsObject>> SQLDatabase::browse(const std::unique_pt
...
 
     std::vector<std::shared_ptr<CdsObject>> result;
-
-    std::unique_ptr<SQLRow> row;
-    while ((row = sqlResult->nextRow())) {
-        auto obj = createObjectFromRow(row);
-        result.push_back(obj);
-        row = nullptr; // clear out content from unique_ptr
+    result.reserve(sqlResult->getNumRows());
+    while (auto row = sqlResult->nextRow()) {
```
Comment (KarlStraussberger) [inline src/database/sql_database.cc:855]: For readability we prefer not to mix assignments with checks and leave optimization to the compiler.

**Review by KarlStraussberger — CHANGES_REQUESTED** [review]

File: src/database/sql_database.cc (hunk 3 above)
Comment (matthiasstraka) [reply src/database/sql_database.cc]: It's now fixed wherever suitable

**Comment by matthiasstraka** [comment]
> Please be aware that there is a PR pending for database code. So some of your changes will have to be reworked.
I'll review the code after the other PR is merged

File: src/database/sql_database.cc [hunk 4]
```diff
@@ -1012,7 +1021,8 @@ std::vector<std::string> SQLDatabase::getMimeTypes()
 
     std::vector<std::string> arr;
     arr.reserve(res->getNumRows());
-    while (auto row = res->nextRow()) {
+    std::unique_ptr<SQLRow> row;
+    while (row = res->nextRow()) {
```
Comment (neheb) [inline src/database/sql_database.cc:1025]: This will definitely cause compiler warnings as extra parentheses are needed.

**Comment by neheb** [comment]
I’d rebase and squash the last two commits into the first as you’re undoing your own changes.

**Comment by matthiasstraka** [comment]
I think the current state of my PR has exceeded my knowledge of git. I will close it and restart with more precise commits (cherry picked by topic) and without the need for merges.

**Pull request closed without merging**